### PR TITLE
ADR for using Click to power dbts CLI

### LINKED ISSUE #6954: [CT-2105] ADR for Click Dependency
### Describe the feature

Parity ticket for CT-877

### PR DESCRIPTION
**Pull request by iknox-fa**
resolves #6954

### Description
ADR for the decision to use Click to power dbts CLI.

### Checklist
- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR

### CODE REVIEW

File: docs/arch/adr-006-click-for-cli.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  * We will ensure the Click app and its commands are importable and invokable by third parties.
+  * We will document the Click app and its commands and provide them as a top-level programmatic interface for automating the usage of dbt.
+ 
+## Status
+ Accepted
+
+## Consequences
+
+ ### Negatives
+  * Click is rather insistent on keeping the parameters from a parent command isolated from the child command. This didn't fit well with the multi-level-ness of the legacy interface and caused us to implement somewhat hacky logic to create a dummy click context and use that as a representation of the parent command. The primary example of this can be seen in `dbt.cli.flags`.
```
Comment (MichelleArk) [inline docs/arch/adr-006-click-for-cli.md:35]: What do you think about adding something in a similar vein in terms of the need to add something like `dbt.cli.main.dbtRunner` for programmatic invocations (for catching cli (and click)-specific exceptions, and providing an interface that handles creating a correctly built click context for programmatic invocations.

Comment (iknox-fa) [reply docs/arch/adr-006-click-for-cli.md:35]: Good call, I'll update shortly.

Comment (iknox-fa) [reply docs/arch/adr-006-click-for-cli.md:35]: Updated!

File: docs/arch/adr-006-click-for-cli.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ ### Research
+  Research was done to discover an alternative as part of an overall API improvement effort. [Click](https://click.palletsprojects.com/) was selected to replace argparse for the following reasons:
+  * Click provides a more robust ecosystem for building CLIs without being overly heavy-weight.
+  * Click is probably the most widely adapted CLI framework for Python, which provides a significant mindshare in the dev community.
+  * Click's composability allows for future code combining various commands into a single programmatic entry point with very little work.
+  * Click provides a plugin ecosystem that allows us to take advantage of third-party contributions.
+
+ ### Alternatives considered
+  * No change
+  * Typer
```
Comment (ChenyuLInx) [inline docs/arch/adr-006-click-for-cli.md:20]: Can we write a bit about why we chose `click` over `typer`? I think either way is valid choice but want to have something here so we can reference in the future.

File: docs/arch/adr-006-click-for-cli.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * `argparse` does not provide any plugin architecture or other methods to add functionality from third parties, leaving us to build any customizations entirely on our own. 
+
+## Decision
+
+ ### Research
+  Research was done to discover an alternative as part of an overall API improvement effort. [Click](https://click.palletsprojects.com/) was selected to replace argparse for the following reasons:
+  * Click provides a more robust ecosystem for building CLIs without being overly heavy-weight.
+  * Click is probably the most widely adapted CLI framework for Python, which provides a significant mindshare in the dev community.
+  * Click's composability allows for future code combining various commands into a single programmatic entry point with very little work.
+  * Click provides a plugin ecosystem that allows us to take advantage of third-party contributions.
```
Comment (ChenyuLInx) [inline docs/arch/adr-006-click-for-cli.md:16]: Do you mean things like [click-plugins](https://github.com/click-contrib/click-plugins)? Can we add some example here so people reading this knows better about what are we thinking about.

**Pull request closed without merging**